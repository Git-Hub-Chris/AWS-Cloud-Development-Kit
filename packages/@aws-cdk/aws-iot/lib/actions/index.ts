export * from './cloudwatch-alarm-action';
export * from './cloudwatch-logs-action';
export * from './cloudwatch-metric-action';
export * from './dynamodb-action';
export * from './dynamodb-v2-action';
export * from './lambda-action';
export * from './republish-action';
export * from './s3-action';
export * from './sns-action';
export * from './sqs-action';
