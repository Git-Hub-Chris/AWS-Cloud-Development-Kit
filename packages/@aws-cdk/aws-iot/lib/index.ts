export * from './action';
export * from './actions';
export * from './topic-rule';
export * from './topic-rule-ref';

// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
