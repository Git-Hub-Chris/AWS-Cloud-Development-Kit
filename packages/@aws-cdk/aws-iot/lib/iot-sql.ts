import { Construct } from 'constructs';

/**
 * The type returned from the `bind()` method in {@link IotSql}.
 */
export interface IotSqlConfig {
  /**
   * The version of the SQL rules engine to use when evaluating the rule.
   */
  readonly awsIotSqlVersion: string;

  /**
   * The SQL statement used to query the topic.
   */
  readonly sql: string;
}

/**
 * Defines AWS IoT SQL.
 */
export abstract class IotSql {
  /**
   * Uses the original SQL version built on 2015-10-08.
   *
   * @param sql The actual SQL-like syntax query
   * @returns Instance of IotSql
   */
  public static fromStringAsVer20151008(sql: string): IotSql {
    return new Ver_2015_10_08_IotSql(sql);
  }

  /**
   * Uses the SQL version built on 2016-03-23.
   *
   * @param sql The actual SQL-like syntax query
   * @returns Instance of IotSql
   */
  public static fromStringAsVer20160323(sql: string): IotSql {
    return new Ver_2016_03_23_IotSql(sql);
  }

  /**
   * Uses the most recent beta SQL version. If you use this version, it might
   * introduce breaking changes to your rules.
   *
   * @param sql The actual SQL-like syntax query
   * @returns Instance of IotSql
   */
  public static fromStringAsVerNewestUnstable(sql: string): IotSql {
    return new VerNewestUnstableIotSql(sql);
  }

  /**
   * Returns the IoT SQL configuration.
   */
  public abstract bind(scope: Construct): IotSqlConfig;
}

class Ver_2015_10_08_IotSql extends IotSql {
  constructor(private readonly sql: string) {
    super();

    validateSqlEmpty(sql);
  }

  bind(_scope: Construct): IotSqlConfig {
    return {
      awsIotSqlVersion: '2015-10-08',
      sql: this.sql,
    };
  }
}

class Ver_2016_03_23_IotSql extends IotSql {
  constructor(private readonly sql: string) {
    super();

    validateSqlEmpty(sql);
  }

  bind(_scope: Construct): IotSqlConfig {
    return {
      awsIotSqlVersion: '2016-03-23',
      sql: this.sql,
    };
  }
}

class VerNewestUnstableIotSql extends IotSql {
  constructor(private readonly sql: string) {
    super();

    validateSqlEmpty(sql);
  }

  bind(_scope: Construct): IotSqlConfig {
    return {
      awsIotSqlVersion: 'beta',
      sql: this.sql,
    };
  }
}

function validateSqlEmpty(sql: string) {
  if (sql === '') {
    throw new Error('sql cannot be empty.');
  }
}
