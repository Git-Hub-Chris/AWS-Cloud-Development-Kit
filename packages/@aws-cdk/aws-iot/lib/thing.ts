import { Resource, IResource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnThing } from './iot.generated';

/**
 * Represents an AWS IoT Thing
 */
export interface IThing extends IResource {
  /**
   * The name of the Thing.
   *
   * @attribute
   */
  readonly thingName: string;
}

/**
 * Properties for defining an AWS IoT Thing
 */
export interface ThingProps {
  /**
   * The name of the Thing.
   *
   * @default - an automatically generated name
   */
  readonly thingName?: string;

  /**
   * A JSON string containing up to three key-value pair in JSON format.
   *
   * @default - None
   */
  readonly attributes?: { [key: string]: string }
}

/**
 * Defines an AWS IoT Thing in this stack.
 */
export class Thing extends Resource {
  /**
   * Import an existing AWS IoT Thing provided an ARN
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param thingName AWS IoT Thing name.
   */
  public static fromThingName(scope: Construct, id: string, thingName: string): IThing {
    const resourceName = thingName;

    class Import extends Resource implements IThing {
      public readonly thingName = resourceName;
    }
    return new Import(scope, id);
  }

  /**
   * The name of the Thing.
   *
   * @attribute
   */
  public readonly thingName: string;

  constructor(scope: Construct, id: string, props?: ThingProps) {
    super(scope, id, { physicalName: props?.thingName });

    if (props?.attributes && Object.keys(props?.attributes).length > 3) {
      throw new Error('Thing cannot have more than three attributes.');
    }

    const resource = new CfnThing(this, 'Resource', {
      thingName: this.physicalName,
      attributePayload: props?.attributes && { attributes: props.attributes },
    });

    this.thingName = this.getResourceNameAttribute(resource.ref);
  }
}
