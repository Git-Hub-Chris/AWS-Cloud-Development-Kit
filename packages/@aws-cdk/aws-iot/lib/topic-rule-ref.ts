import { IResource } from '@aws-cdk/core';

/**
 * Represents an AWS IoT Rule
 */
export interface ITopicRule extends IResource {
  /**
   * The value of the topic rule Amazon Resource Name (ARN), such as
   * arn:aws:iot:us-east-2:123456789012:rule/rule_name
   *
   * @attribute
   */
  readonly topicRuleArn: string;

  /**
   * The name topic rule
   *
   * @attribute
   */
  readonly topicRuleName: string;
}
