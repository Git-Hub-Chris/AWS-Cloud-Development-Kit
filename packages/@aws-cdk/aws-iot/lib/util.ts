export { generatePolicyName } from '@aws-cdk/aws-iam/lib/util';
import { Stack } from '@aws-cdk/core';
import { IConstruct } from 'constructs';
import { CertificateAttributes } from './certificate';

export function parsePolicyArn(construct: IConstruct, policyName: string): string {
  return Stack.of(construct).formatArn({
    account: Stack.of(construct).account,
    region: Stack.of(construct).region,
    service: 'iot',
    resource: 'policy',
    resourceName: policyName,
  });
}

export function parseCertificateArn(construct: IConstruct, props: CertificateAttributes): string {
  if (props.certificateArn) {
    return props.certificateArn;
  }

  if (props.certificateId) {
    return Stack.of(construct).formatArn({
      region: Stack.of(construct).region,
      account: Stack.of(construct).account,
      service: 'iot',
      resource: 'cert',
      resourceName: props.certificateId,
    });
  }

  throw new Error('Cannot determine certifiate ARN. At least `certificateArn` or `certifiateId` is needed');
}

export function parseCertificateId(construct: IConstruct, props: CertificateAttributes): string {
  if (props.certificateId) {
    return props.certificateId;
  }

  // extract certificate id from certificate arn
  if (props.certificateArn) {
    const resourceName = Stack.of(construct).parseArn(props.certificateArn).resourceName || 'cert/';
    return resourceName.replace('cert/', '');
  }

  throw new Error('Cannot determine certifiate ID. At least `certificateArn` or `certifiateId` is needed');
}
