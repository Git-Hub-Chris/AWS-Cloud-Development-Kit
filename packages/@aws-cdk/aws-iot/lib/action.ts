import { CfnTopicRule } from './iot.generated';
import { ITopicRule } from './topic-rule-ref';

/**
 * An abstract action for TopicRule.
 */
export interface IAction {
  /**
   * Returns the topic rule action specification.
   *
   * @param rule The TopicRule that would trigger this action.
   */
  bind(rule: ITopicRule): ActionConfig;
}

/**
 * Properties for an topic rule action
 */
export interface ActionConfig {
  /**
   * An action to set state of an Amazon CloudWatch alarm.
   * @default None
   */
  readonly cloudwatchAlarm?: CfnTopicRule.CloudwatchAlarmActionProperty;
  /**
   * An action to send data to Amazon CloudWatch Logs.
   * @default None
   */
  readonly cloudwatchLogs?: CfnTopicRule.CloudwatchLogsActionProperty;
  /**
   * An action to capture an Amazon CloudWatch metric.
   * @default None
   */
  readonly cloudwatchMetric?: CfnTopicRule.CloudwatchMetricActionProperty;
  /**
   * An action to write all or part of an MQTT message to an Amazon DynamoDB table.
   * @default None
   */
  readonly dynamoDb?: CfnTopicRule.DynamoDBActionProperty;
  /**
   * An action to write all or part of an MQTT message to an Amazon DynamoDB table.
   * @default None
   */
  readonly dynamoDBv2?: CfnTopicRule.DynamoDBv2ActionProperty;
  /**
   * An action to write data from MQTT messages to an Amazon OpenSearch Service domain.
   * @default None
   */
  readonly elasticsearch?: CfnTopicRule.ElasticsearchActionProperty;
  /**
   * An action to send data from an MQTT message to an Amazon Kinesis Data Firehose stream.
   * @default None
   */
  readonly firehose?: CfnTopicRule.FirehoseActionProperty;
  /**
   * An action to send data from an MQTT message to a web application or service.
   * @default None
   */
  readonly http?: CfnTopicRule.HttpActionProperty;
  /**
   * An action to send data from an MQTT message to an AWS IoT Analytics channel.
   * @default None
   */
  readonly iotAnalytics?: CfnTopicRule.IotAnalyticsActionProperty;
  /**
   * An action to send data from an MQTT message to an AWS IoT Events input.
   * @default None
   */
  readonly iotEvents?: CfnTopicRule.IotEventsActionProperty;
  /**
   * An action to send data from an MQTT message to asset properties in AWS IoT SiteWise.
   * @default None
   */
  readonly iotSiteWise?: CfnTopicRule.IotSiteWiseActionProperty;
  /**
   * An action to sends messages directly to your Amazon MSK or self-managed Apache Kafka clusters for data analysis and visualization.
   * @default None
   */
  readonly kafka?: CfnTopicRule.KafkaActionProperty;
  /**
   * An action to write data from an MQTT message to Amazon Kinesis Data Streams.
   * @default None
   */
  readonly kinesis?: CfnTopicRule.KinesisActionProperty;
  /**
   * An action to invoke an AWS Lambda function, passing in an MQTT message.
   * @default None
   */
  readonly lambda?: CfnTopicRule.LambdaActionProperty;
  /**
   * An action to republish an MQTT message to another MQTT topic.
   * @default None
   */
  readonly republish?: CfnTopicRule.RepublishActionProperty;
  /**
   * An action to write the data from an MQTT message to an Amazon S3 bucket.
   * @default None
   */
  readonly s3?: CfnTopicRule.S3ActionProperty;
  /**
   * An action to send the data from an MQTT message as an Amazon SNS push notification.
   * @default None
   */
  readonly sns?: CfnTopicRule.SnsActionProperty;
  /**
   * An action to send data from an MQTT message to an Amazon SQS queue.
   * @default None
   */
  readonly sqs?: CfnTopicRule.SqsActionProperty;
  /**
   * An action to start an AWS Step Functions state machine.
   * @default None
   */
  readonly stepFunctions?: CfnTopicRule.StepFunctionsActionProperty;
  /**
   * An action to write attributes (measures) from an MQTT message into an Amazon Timestream table.
   * @default None
   */
  readonly timestream?: CfnTopicRule.TimestreamActionProperty;
}
