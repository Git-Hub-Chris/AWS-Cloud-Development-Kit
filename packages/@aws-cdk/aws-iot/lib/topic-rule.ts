import { ArnFormat, Lazy, Resource, Stack } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { IAction } from './action';
import { CfnTopicRule } from './iot.generated';
import { ITopicRule } from './topic-rule-ref';

/**
 * The version of the SQL rules engine to use when evaluating the rule.
 *
 * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-rule-sql-version.html
 */
export enum AwsIotSqlVersion {
  /**
   * 2015-10-08 – The original SQL version built on 2015-10-08.
   */
  VER_2015_10_08 = '2015-10-08',
  /**
   * 2016-03-23 – The SQL version built on 2016-03-23.
   */
  SQL_2016_03_23 = '2016-03-23',
  /**
   * beta – The most recent beta SQL version. If you use this version, it might introduce breaking changes to your rules.
   */
  SQL_BETA = 'beta',
}

/**
 * Properties for defining an AWS IoT Rule
 */
export interface TopicRuleProps {
  /**
   * The name of the rule.
   * @default None
   */
  readonly topicRuleName?: string;
  /**
   * The rule payload.
   */
  readonly topicRulePayload: TopicRulePayloadProperty;
}

/**
 * Properties for defining details of an AWS IoT Rule
 */
export interface TopicRulePayloadProperty {
  /**
   * The actions associated with the rule.
   *
   * @default No actions will be perform
   */
  readonly actions?: Array<IAction>;
  /**
   * The version of the SQL rules engine to use when evaluating the rule.
   *
   * @default AwsIotSqlVersion.SQL_2015_10_08
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-create-rule.html
   */
  readonly awsIotSqlVersion?: AwsIotSqlVersion;
  /**
   * A textual description of the rule.
   *
   * @default None
   */
  readonly description?: string;
  /**
   * The action AWS IoT performs when it is unable to perform a rule's action.
   *
   * @default No action will be perform
   */
  readonly errorAction?: IAction;
  /**
   * Specifies whether the rule is disabled.
   *
   * @default false
   */
  readonly ruleDisabled?: boolean;
  /**
   * A simplified SQL syntax to filter messages received on an MQTT topic and push the data elsewhere.
   *
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-sql-reference.html
   */
  readonly sql: string;
}

/**
 * Defines an AWS IoT Rule in this stack.
 */
export class TopicRule extends Resource implements ITopicRule {
  /**
   * Import an existing AWS IoT Rule provided an ARN
   *
   * @param scope The parent creating construct (usually `this`).
   * @param id The construct's name.
   * @param topicRuleArn AWS IoT Rule ARN (i.e. arn:aws:iot:<region>:<account-id>:rule/MyRule).
   */
  public static fromTopicRuleArn(scope: Construct, id: string, topicRuleArn: string): ITopicRule {
    const parts = Stack.of(scope).splitArn(topicRuleArn, ArnFormat.SLASH_RESOURCE_NAME);
    if (!parts.resourceName) {
      throw new Error('Invalid topic rule arn: topicRuleArn has no resource name.');
    }
    const resourceName = parts.resourceName;

    class Import extends Resource implements ITopicRule {
      public readonly topicRuleArn = topicRuleArn;
      public readonly topicRuleName = resourceName;
    }
    return new Import(scope, id);
  }

  /**
   * Arn of this rule
   * @attribute
   */
  public readonly topicRuleArn: string;
  /**
   * Name of this rule
   * @attribute
   */
  public readonly topicRuleName: string;

  private readonly actions = new Array<CfnTopicRule.ActionProperty>();

  constructor(scope: Construct, id: string, props: TopicRuleProps) {
    super(scope, id, {
      physicalName: props.topicRuleName,
    });

    if (props.topicRulePayload.sql === '') {
      throw new Error('\'topicRulePayload.sql\' cannot be empty.');
    }

    const resource = new CfnTopicRule(this, 'Resource', {
      ruleName: this.physicalName,
      topicRulePayload: {
        actions: Lazy.any({ produce: () => this.actions }),
        awsIotSqlVersion: props.topicRulePayload.awsIotSqlVersion ?? AwsIotSqlVersion.SQL_2015_10_08,
        description: props.topicRulePayload.description,
        errorAction: props.topicRulePayload.errorAction?.bind(this),
        ruleDisabled: props.topicRulePayload.ruleDisabled ?? false,
        sql: props.topicRulePayload.sql,
      },
    });

    this.topicRuleArn = this.getResourceArnAttribute(resource.attrArn, {
      service: 'iot',
      resource: 'rule',
      resourceName: this.physicalName,
    });
    this.topicRuleName = this.getResourceNameAttribute(resource.ref);

    props.topicRulePayload.actions?.forEach(action => {
      this.addAction(action);
    });
  }

  /**
   * Add a action to the rule.
   *
   * @param action the action to associate with the rule.
   */
  public addAction(action: IAction): void {
    const actionConfig = action.bind(this);

    const keys = Object.keys(actionConfig);
    if (keys.length === 0) {
      throw new Error('Empty actions are not allowed. Please define one type of action');
    }
    if (keys.length >= 2) {
      throw new Error(`Each object in the actions list can only have one action defined. keys: ${keys}`);
    }

    this.actions.push(actionConfig);
  }
}
