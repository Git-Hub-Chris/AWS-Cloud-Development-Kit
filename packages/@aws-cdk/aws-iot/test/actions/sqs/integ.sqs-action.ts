/// !cdk-integ pragma:ignore-assets
import * as logs from '@aws-cdk/aws-logs';
import * as sqs from '@aws-cdk/aws-sqs';
import * as cdk from '@aws-cdk/core';
import * as iot from '../../../lib';

const app = new cdk.App();

class TestStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const topicRule = new iot.TopicRule(this, 'TopicRule', {
      topicRulePayload: {
        sql: "SELECT topic(2) as device_id FROM 'device/+/data'",
        errorAction: new iot.CloudwatchLogsAction(new logs.LogGroup(this, 'logs')),
      },
    });

    const queue = new sqs.Queue(this, 'MyQueue', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    topicRule.addAction(new iot.SqsAction(queue));
  }
}

new TestStack(app, 'test-stack');
app.synth();
