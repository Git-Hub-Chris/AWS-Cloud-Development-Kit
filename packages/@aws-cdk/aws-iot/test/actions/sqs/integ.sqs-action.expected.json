{
  "Resources": {
    "logs0B6081B1": {
      "Type": "AWS::Logs::LogGroup",
      "Properties": {
        "RetentionInDays": 731
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "TopicRuleTopicRuleActionRole246C4F77": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "iot.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "TopicRuleTopicRuleActionRoleDefaultPolicy99ADD687": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:DescribeLogStreams",
                "logs:PutLogEvents"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "logs0B6081B1",
                  "Arn"
                ]
              }
            },
            {
              "Action": "sqs:SendMessage",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyQueueE6CA6235",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "TopicRuleTopicRuleActionRoleDefaultPolicy99ADD687",
        "Roles": [
          {
            "Ref": "TopicRuleTopicRuleActionRole246C4F77"
          }
        ]
      }
    },
    "TopicRule40A4EA44": {
      "Type": "AWS::IoT::TopicRule",
      "Properties": {
        "TopicRulePayload": {
          "Actions": [
            {
              "Sqs": {
                "QueueUrl": {
                  "Ref": "MyQueueE6CA6235"
                },
                "RoleArn": {
                  "Fn::GetAtt": [
                    "TopicRuleTopicRuleActionRole246C4F77",
                    "Arn"
                  ]
                },
                "UseBase64": false
              }
            }
          ],
          "AwsIotSqlVersion": "2015-10-08",
          "ErrorAction": {
            "CloudwatchLogs": {
              "LogGroupName": {
                "Ref": "logs0B6081B1"
              },
              "RoleArn": {
                "Fn::GetAtt": [
                  "TopicRuleTopicRuleActionRole246C4F77",
                  "Arn"
                ]
              }
            }
          },
          "RuleDisabled": false,
          "Sql": "SELECT topic(2) as device_id FROM 'device/+/data'"
        }
      }
    },
    "MyQueueE6CA6235": {
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    }
  }
}